src/cpp/src/speculative_decoding/speculative_decoding_stateful_eagle3.cpp

```diff
@@ -0,0 +1,1242 @@
+    m_raw_perf_metrics.m_durations.clear();
+    m_raw_perf_metrics.m_batch_sizes.clear();
+
+    log_debug(LogCategory::INIT, "State reset");
+}
+
+void Eagle3InferWrapperBase::release_memory() {
+    m_request.get_compiled_model().release_memory();
+}
+
+void Eagle3InferWrapperBase::build_model_inputs(size_t input_token_count,
+                                                ov::Tensor& input_ids,
+                                                ov::Tensor& attention_mask,
+                                                ov::Tensor& position_ids) {
+    auto current_sequence = get_current_sequence();
+    OPENVINO_ASSERT(current_sequence, "SequenceGroup not initialized");
+
+    const auto& prompt_ids = m_sequence_group->get_prompt_ids();
+    const auto& generated_ids = current_sequence->get_generated_ids();
+
+    const size_t prompt_len = prompt_ids.size();
+    const size_t generated_len = generated_ids.size();
+    const size_t total_len = prompt_len + generated_len;
+    const size_t start_pos = total_len - input_token_count;
+
+    OPENVINO_ASSERT(input_token_count > 0 && input_token_count <= total_len,
+                    "Invalid input_token_count: ",
+                    input_token_count,
+                    ", total_len: ",
+                    total_len);
+
+    // Allocate tensors
+    input_ids = ov::Tensor(ov::element::i64, {1, input_token_count});
+    position_ids = ov::Tensor(ov::element::i64, {1, input_token_count});
+
+    int64_t* input_ids_ptr = input_ids.data<int64_t>();
+    int64_t* position_ids_ptr = position_ids.data<int64_t>();
+
+    // Fill input_ids and position_ids from sequence
+    if (start_pos < prompt_len) {
+        // Part from prompt
+        const size_t prompt_count = std::min(input_token_count, prompt_len - start_pos);
+        std::memcpy(input_ids_ptr, prompt_ids.data() + start_pos, prompt_count * sizeof(int64_t));
+        for (size_t i = 0; i < prompt_count; ++i) {
+            position_ids_ptr[i] = static_cast<int64_t>(start_pos + i);
+        }
+
+        // Part from generated (if any)
+        if (input_token_count > prompt_count) {
+            const size_t generated_count = input_token_count - prompt_count;
+            std::memcpy(input_ids_ptr + prompt_count, generated_ids.data(), generated_count * sizeof(int64_t));
+            for (size_t i = 0; i < generated_count; ++i) {
+                position_ids_ptr[prompt_count + i] = static_cast<int64_t>(prompt_len + i);
+            }
+        }
+    } else {
+        // All from generated
+        const size_t generated_start = start_pos - prompt_len;
+        std::memcpy(input_ids_ptr, generated_ids.data() + generated_start, input_token_count * sizeof(int64_t));
+        for (size_t i = 0; i < input_token_count; ++i) {
+            position_ids_ptr[i] = static_cast<int64_t>(prompt_len + generated_start + i);
+        }
+    }
+
+    // Build attention mask (all ones up to last position + 1)
```

Comment:
Could  you please add a comment that `last position + 1` is needed to cover newly generated token, if I am right about this?